### PR TITLE
Add ViT SAC encoder and policy support

### PR DESCRIPTION
**Pull request by robkeeper**

### CODE REVIEW

**Comment by JaimeParker** [comment]
**refactor: 用 encoder capability declaration 取代 ViTSACPolicy** (`a2d7b54`)

这次提交解决了 PR 原始实现中"为每种 encoder 单独派生 Policy 子类"的问题。核心变化是将 ViT 的特殊能力声明从 Policy 层下移到 Encoder 层：

- 删除 `vit_sac_policy.py`，ViT 不再需要独立的 Policy 类。
- `BaseFeaturesExtractor` 新增两个默认正确的钩子：`structured_feature_config()`（返回 `None`）和 `prepare_batch()`（no-op）；非 ViT encoder 零改动。
- `ViTCombinedExtractor` 实现 `structured_feature_config()` 返回 `TokenAndPropFeatureConfig`，声明 `[tokens_flat, prop]` 的 flat layout；`prepare_batch()` 负责带 augmentation 的特征缓存。
- `SACPolicy` 读取 `structured_feature_config()` 自适应选择 `SpatialEmbQEnsemble`（ViT）或 `EnsembleQCritic`（flat），以及是否构建 `_TokenCompressor` actor adapter；不再有任何 `isinstance` 分支。
- `SACCore.train/train_high_utd` 无条件调用 `features_extractor.prepare_batch()`，取代原来的 `getattr(self.policy, "prepare_train_batch", None)` anti-pattern。
- `SpatialEmbQHead/SpatialEmbQEnsemble` 移入新文件 `rl_garden/networks/spatial_critic.py`，与任何 encoder 类型完全解耦。
- `VitImageEncoder` → `ViTImageEncoder`；CLI 函数 `sac_family_policy_kwargs_from_args` → `vit_policy_kwargs_from_args`，extractor kwargs 不再携带 policy-head 超参。

**Comment by JaimeParker** [comment]
**fix: 补全 CQL/CalQL/WSRL 的 structured extractor 链路** (`ef0ef25`)

上一个 commit 只重构了 SAC 和 ResidualSAC 路径；这次 commit 补齐了遗漏部分：

- **重命名**：`ViTCombinedExtractor` → `ViTTokenAndPropExtractor`，名字更准确地反映输出契约（`[tokens_flat, prop]`），与通用的 `CombinedExtractor`（dict → flat）在命名上明确区分。
- **CQL actor bypass fix**：原实现的 `_sample_n_actions_with_log_probs` 直接访问 `self.policy.actor(features)`，绕过了 `_TokenCompressor`。新增 `SACPolicy._sample_actor_actions(features, n)` 方法（先走 `_transform_features_for_actor`，再采样），CQL 改为调用该方法，解决了 ViT + `actor_feature_dim` 路径下的维度不匹配问题。
- **参数链路**：`actor_feature_dim`/`critic_spatial_emb_dim` 补入 `CQLCore`、`CQL`、`CalQL`、`WSRL` 的构造函数并转发到 `SACPolicy`，并写入 checkpoint metadata。
- **`_resolve_policy_kwargs` 修复**：CQL 的 policy kwargs 解析逻辑原来会在检测到自定义 extractor class 时仍构造默认 extractor kwargs，导致 `ViTTokenAndPropExtractor` 的 kwargs 被污染或覆盖；现在统一以 `has_custom_class` 为分支条件。
- **入口修复**：`pretrain_offline.py` 的离线 CQL 路径补入 ViT extractor 注入；`generate_wsrl_dataset.py` 的 `vit_actor_feature_dim` 默认值从 `128` 改为 `None`，避免 plain_conv 路径触发 `ValueError`。

**Comment by JaimeParker** [comment]
@robkeeper 麻烦根据上面两次更新后的代码风格，重新跑一下这个 PR 的测试。

这两次 commit 的定位是纯架构重构，不涉及算法本身的改动：ViT encoder 的前向逻辑、augmentation 策略、SpatialEmbQHead 的 Q 计算均未变动，只是把它们从 Policy 层挪到了正确的位置。理论上训练曲线不应有任何偏移，如果测试结果发现性能差异，请在这里贴出来，我们一起排查。

**Comment by JaimeParker** [comment]
## fix(encoders): restore vit_actor_feature_dim default, image norm by obs space

**Commit:** 7444cb3（squash of be688ff + 2d8117b）

---

### 1. `vit_actor_feature_dim` 默认值回退（训练回归根因）

`cli_args.py` 中 `VisionArgs` 和 `OfflineVisionArgs` 各一处：

```python
# 重构后（引入回归）
vit_actor_feature_dim: Optional[int] = None

# 修复后
vit_actor_feature_dim: int = 128
```

`None` 时 `SACPolicy` 跳过 `_TokenCompressor`，actor 输入从 ~137 维膨胀为 ~6281 维。WandB `compare_runs_tool` 对比 seed=2 前后两条 run（`mefyz8zw` vs `w00b2i27`），55 项超参数完全一致，唯此项不同，与所有指标退化吻合。

---

### 2. 图像归一化判据改为 observation space bounds

`encoders/base.py` 新增 `image_needs_normalization(space)`：

```python
return space.dtype == np.uint8 or float(space.high.max()) > 1.0
```

`ViTTokenAndPropExtractor` 和 `CombinedExtractor` 在 `__init__` 预计算 `_needs_norm: frozenset[str]`，`_prepare_image` 改用 `key in self._needs_norm`。移除了 `dtype == uint8 or key == "rgb"` 等依赖 tensor dtype 或 key 名称的启发式判断，正确处理 uint8、float32[0,255]、float32[0,1] 三种输入源。

---

### 3. `ViTImageEncoder.forward_augment` 独立字段

```python
# 新增参数，默认 False（对齐重构前行为）
def __init__(self, ..., forward_augment: bool = False):
    self.forward_augment = forward_augment

def forward(self, x):
    tokens = self.encode_tokens(x, augment=self.forward_augment and self.training)
```

用独立字段 `self.forward_augment` 替代 `self.training`，与 `nn.Module` 的 train/eval 传播解耦，调用方可显式 opt-in。默认 `False` 对齐原始行为；`ViTTokenAndPropExtractor` 已通过 `prepare_batch()` 在 batch 级增强，不受影响。

---

**验证**：76 tests passed。
